### ISSUE TITLE
adapter.createEach not called

### ISSUE THREAD

**Issue by atiertant** [balderdashy/waterline#1007]
adapter.createEach is only called in alter mode to reinsert data.
shouldn't it optimize insertion of multiple data ?

**Comment by dmarcelino** [comment]
Yes, it should be called when calling `.create([], cb);`. This looks like a bug.

**Comment by devinivy** [comment]
I haven't tested, but I'm afraid `createEach` is possibly never optimized.

See `Collection#createEach` always passing its work onto `Collection#create`: https://github.com/balderdashy/waterline/blob/fe402789bd89d343c3f673c95e01415a65ce8176/lib/waterline/query/aggregate.js#L24

On the other hand, the adapter `createEach` method is ready to optimize, but I don't think the adapter method is ever used: https://github.com/balderdashy/waterline/blob/fe402789bd89d343c3f673c95e01415a65ce8176/lib/waterline/adapter/aggregateQueries.js#L13

It appears `Collection#create` with an array does ask `Collection#createEach` to do the work, so this issue probably appears when using either syntax: https://github.com/balderdashy/waterline/blob/fe402789bd89d343c3f673c95e01415a65ce8176/lib/waterline/query/dql/create.js#L60-L63

If this is indeed a bug, we'll have to do some work to hook the lifecycle into `createEach` when it's optimized by an adapter.

**Comment by dmarcelino** [comment]
> I haven't tested, but I'm afraid `createEach` is possibly never optimised.

Back in https://github.com/balderdashy/waterline/pull/980#issuecomment-98466753 I did notice sails-postgresql and sails-mysql were using waterline's `.createEach()` shim, even though they define their own `.createEach()` in [adapter.js#L402](https://github.com/balderdashy/sails-postgresql/blob/master/lib/adapter.js#L402) and [adapter.js#L424](https://github.com/balderdashy/sails-mysql/blob/master/lib/adapter.js#L424) respectively.

> If this is indeed a bug,

I think it is.

> we'll have to do some work to hook the lifecycle into createEach when it's optimized by an adapter.

Agreed.

**Comment by masumsoft** [comment]
Same here. It is causing really slow write performance as batch writes are not using adapter's createEach() method, rather sending individual create queries and thus insert performance is drastically degraded.

**Comment by dmarcelino** [comment]
Hi @masumsoft, I've left some comments (https://github.com/balderdashy/waterline/pull/1018#issuecomment-104279274) on PR #1018. Please take a look.

**Comment by masumsoft** [comment]
Hi @dmarcelino, Thanks for the nice feedback. I was a bit busy this week, didn't have much time to look at it, will try to resolve the issues mentioned by you asap.

**Comment by masumsoft** [comment]
Updated the PR #1018 with new commits, details available there.

**Comment by sailsbot** [comment]
Thanks for posting, @atiertant.  I'm a repo bot-- nice to meet you!

It has been 60 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by devinivy** [comment]
This has a PR in the works.

**Comment by particlebanana** [comment]
The main issue with this is handling nested creates. That's why it currently loops through and runs `.create` for each item in the array. So each item needs to essentially go through this flow:

• Set model default values
• Create any nested records that belong to the parent
• Run any `beforeCreate` lifecycle callbacks that have been defined
• Create the record
• Associate any nested m:m records with the newly created record
• Run any `afterCreate` callbacks that have been defined
• Turn object into a Waterline model

If we were to pass that whole array to the adapter's `createEach` method we would need to flatten out the nested records and handle the logic for each nested collection item. Which I don't think you could get in a single insert query anyway because they would be inserting records into multiple tables.

Now we could simplify this and say you can't do nested creates in a `createEach` which would allow this to be much simpler. Then you just loop through the array running beforeCreate and adding default values then pass the array to the adapter.

**Comment by atiertant** [comment]
@particlebanana an other way would be to group flat insert by model so an array of two level nested create could be insert in two createEach like deepPopulate but with nested create logic.

**Comment by sailsbot** [comment]
Thanks for posting, @atiertant.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by particlebanana** [comment]
This is on the roadmap file. Closing it now. To clarify there needs to be a proposal as to how this change effects all the various nested updates operations.
